Tools/cases_generator/generate_cases.py

```diff
@@ -83,28 +111,43 @@ def block(self, head: str):
             yield
         self.emit("}")
 
-    def stack_adjust(self, diff: int):
+    def stack_adjust(self, diff: int, input_effects: list[StackEffect], output_effects: list[StackEffect]):
+        # TODO: Get rid of 'diff' parameter
+        shrink, isym = list_effect_size(input_effects)
+        grow, osym = list_effect_size(output_effects)
+        diff += grow - shrink
+        if isym and isym != osym:
+            self.emit(f"STACK_SHRINK({isym});")
+        if diff < 0:
+            self.emit(f"STACK_SHRINK({-diff});")
         if diff > 0:
             self.emit(f"STACK_GROW({diff});")
-        elif diff < 0:
-            self.emit(f"STACK_SHRINK({-diff});")
+        if osym and osym != isym:
+            self.emit(f"STACK_GROW({osym});")
```

Comment:
See my comment earlier. It probably doesn't matter much, but it *feels* like we should be combining these into a single `STACK_GROW`/`STACK_SHRINK` per instruction.

Reply:
I thought about that. There are some problematic theoretical cases where the generator would need to understand the symbolic sizes to be able to tell if it's a net grow or shrink operation, and it can't first do a grow/shrink of the numeric part only, since you could have e.g.
```
inst(FOO, (a, b[oparg] -- c, d))
```
which shrinks if `oparg == 0` but grows for `oparg >= 2`. We cannot safely do this `as grow(1); shrink(oparg)`  or `shrink(oparg); grow(1)`, nor as `grow(1 - oparg)` nor as `shrink(oparg - 1)`. It must be done as `shrink(1 + oparg); grow(2)`, or use `STACK_ADJUST()`, but the latter doesn't do bounds checking in debug mode. The best way is `shrink(oparg); grow(1)`, but I left it even simpler, as `shrink(1); shrink(oparg); grow(2)`.

Honestly, the case analysis got hairy enough that I decided that this was okay to leave to future generations; the compiler will probably do a fine job optimizing `shrink(a); shrink(b)` to `shrink(a + b)`.